src/Adapters/Auth/oauth2.js

```diff
@@ -0,0 +1,231 @@
+ *      value against the set of appIds in the adapter config. The concept of
+ *      appIds comes from the two major social login providers
+ *      (Google and Facebook). They have not yet implemented the token
+ *      introspection endpoint, but the concept can be valid for any OAuth2
+ *      provider.
+ *      Default: - (undefined)
+ *
+ * 4. "appIds" (array of strings, optional)
+ *      A set of appIds that are used to restrict accepted access tokens based
+ *      on a specific field's value in the token introspection response.
+ *      Default: - (undefined)
+ *
+ * 5. "authorizationHeader" (string, optional)
+ *      The value of the "Authorization" HTTP header in requests sent to the
+ *      introspection endpoint.
+ *      Eg. "Basic dXNlcm5hbWU6cGFzc3dvcmQ="
+ *
+ * 6. "debug" (boolean, optional)
+ *      Enables extensive logging using the "verbose" level.
+ *
+ * The adapter expects requests with the following authData JSON:
+ *
+ * {
+ *   "oauth2": {
+ *     "id": "user's OAuth2 provider-specific id as a string",
+ *     "access_token": "an authorized OAuth2 access token for the user",
+ *   }
+ * }
+ */
+
+import logger from '../../logger';
+const Https = require('https');
+const Parse = require('parse/node').Parse;
+const Path = require('path');
+const Url = require('url');
+const Querystring = require('querystring');
+
+const scriptName = Path.basename(__filename, '.js');
+
+// Returns a promise that fulfills if this user id is valid.
+function validateAuthData(authData, options) {
+  const _logger = _getLogger(validateAuthData.name);
+  const loggingEnabled = _isLoggingEnabled(options);
+  if (loggingEnabled) {
+    _logger.verbose('authData = %s, options = %s',  _objectToString(authData), _objectToString(options));
+  }
+  return requestJson(options, authData.access_token, loggingEnabled).then((response) => {
+    if (response && response.active && (!options || !options.hasOwnProperty('useridField') || !options.useridField || authData.id == response[options.useridField])) {
+      return Promise.resolve();
+    }
+    return Promise.reject(new Parse.Error(
+      Parse.Error.OBJECT_NOT_FOUND,
+      'OAuth2 access token is invalid for this user.'));
+  });
+}
+
+function validateAppId(appIds, authData, options) {
+  const _logger = _getLogger(validateAppId.name);
+  const loggingEnabled = _isLoggingEnabled(options);
+  if (loggingEnabled) {
+    _logger.verbose('appIds = %s, authData = %s, options = %s', _objectToString(appIds), _objectToString(authData), _objectToString(options));
+  }
+  if (options && options.hasOwnProperty('appidField') && options.appidField) {
+    if (!appIds.length) {
+      return Promise.reject(new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'OAuth2 configuration is missing the client app IDs ("appIds" config parameter).'));
+    }
+    return requestJson(options, authData.access_token, loggingEnabled).then((response) => {
+      const appidField = options.appidField;
```

Comment:
can we get a few comments on this logic please?
I have no idea how to debug / maintain this